src/hotspot/share/opto/memnode.cpp

```diff
@@ -3281,25 +3286,38 @@ Node* MergePrimitiveArrayStores::make_merged_input_value(const Node_List& merge_
     jlong mask = (((jlong)1) << bits_per_store) - 1;
     for (uint i = 0; i < merge_list.size(); i++) {
       jlong con_i = merge_list.at(i)->in(MemNode::ValueIn)->get_int();
+#ifdef VM_LITTLE_ENDIAN
       con = con << bits_per_store;
       con = con | (mask & con_i);
+#else // VM_LITTLE_ENDIAN
+      con_i = (mask & con_i) << (i * bits_per_store);
+      con = con | con_i;
+#endif // VM_LITTLE_ENDIAN
     }
     merged_input_value = _phase->longcon(con);
   } else {
     // Pattern: [base >> 24, base >> 16, base >> 8, base] -> base
     //             |                                  |
     //           _store                             first
     //
-    merged_input_value = first->in(MemNode::ValueIn);
-    Node const* base_last;
-    jint shift_last;
-    bool is_true = is_con_RShift(_store->in(MemNode::ValueIn), base_last, shift_last);
+#ifdef VM_LITTLE_ENDIAN
+    Node* hi = _store->in(MemNode::ValueIn);
+    Node* lo = first->in(MemNode::ValueIn);
+#else // VM_LITTLE_ENDIAN
+    // `_store` and `first` are swapped in the diagram above
+    Node* hi = first->in(MemNode::ValueIn);
+    Node* lo = _store->in(MemNode::ValueIn);
+#endif // VM_LITTLE_ENDIAN
```

Comment:
A `swap` could be more concise. But I leave that up to you ;)

Reply:
You're right. It's better to just swap `hi` with `lo` and it matches the comment.